src/CPUSolver.cpp

```diff
@@ -355,111 +262,184 @@ void CPUSolver::normalizeFluxes() {
+                                  * _scalar_flux(r,g);
+      scatter_source = pairwise_sum<FP_PRECISION>(&scatter_sources(tid,0),
                                                 _num_groups);
 
-      /* Set the fission source for FSR r in group G */
-      fsr_fission_source += fission_source * chi[G];
+      _reduced_sources(r,G) = fission_source * chi[G];
+      _reduced_sources(r,G) += scatter_source + _fixed_sources(r,G);
+      _reduced_sources(r,G) *= ONE_OVER_FOUR_PI / sigma_t[G];
+    }
+  }
+
+  delete [] fission_sources;
+  delete [] scatter_sources;
+}
+
 
-      _reduced_sources(r,G) = (fission_source * chi[G] + scatter_source) *
-                              ONE_OVER_FOUR_PI / sigma_t[G];
+/**
+ * @brief Computes the residual between source/flux iterations.
+ * @param res_type the type of residuals to compute 
+ *        (SCALAR_FLUX, FISSION_SOURCE, TOTAL_SOURCE)
+ * @return the average residual in each FSR
+ */
+double CPUSolver::computeResidual(residualType res_type) {
+
+  int norm;
+  double residual;
+  double* residuals = new double[_num_FSRs];
+  memset(residuals, 0., _num_FSRs * sizeof(double));
+
+  if (res_type == SCALAR_FLUX) {
+
+    norm = _num_FSRs * _num_groups;
+
+    for (int r=0; r < _num_FSRs; r++) {
+      for (int e=0; e < _num_groups; e++)
+        if (_old_scalar_flux(r,e) > 0.) {
+          residuals[r] += pow((_scalar_flux(r,e) - _old_scalar_flux(r,e)) / 
+                               _old_scalar_flux(r,e), 2);
+      }
     }
+  }
+
+  else if (res_type == FISSION_SOURCE) {
+
+    norm = _num_fissionable_FSRs * _num_groups;
+
+    double new_fission_source, old_fission_source;
+    FP_PRECISION* nu_sigma_f;
+    Material* material;
 
-    /* Compute the norm of residual of the source in the FSR */
-    if (fsr_fission_source > 0.0)
-      _source_residuals[r] = pow((fsr_fission_source - _old_fission_sources[r])
-                                 / fsr_fission_source, 2);
+    for (int r=0; r < _num_FSRs; r++) {
+      new_fission_source = 0.;
+      old_fission_source = 0.;
+      material = _FSR_materials[r];
+
+      if (material->isFissionable()) {
+        nu_sigma_f = material->getNuSigmaF();
+
+        for (int e=0; e < _num_groups; e++) {
+          new_fission_source += _scalar_flux(r,e) * nu_sigma_f[e];
+          old_fission_source += _old_scalar_flux(r,e) * nu_sigma_f[e];
+        }
 
-    /* Update the old source */
-    _old_fission_sources[r] = fsr_fission_source;
+        if (old_fission_source > 0.)
+          residuals[r] = pow((new_fission_source -  old_fission_source) /
+                              old_fission_source, 2);
+      }
+    }
   }
 
-  /* Sum up the residuals from each FSR */
-  source_residual = pairwise_sum<FP_PRECISION>(_source_residuals, _num_FSRs);
-  source_residual = sqrt(source_residual \
-                         / (_num_fissionable_FSRs * _num_groups));
+  else if (res_type == TOTAL_SOURCE) {
+
+    norm = _num_FSRs * _num_groups;
```

Comment:
No, I was just keeping to the existing normalization constant. But I realize that this is erroneous since the groups are integrated out in the residual calculation. I'll revise the norm for all convergence criteria momentarily.